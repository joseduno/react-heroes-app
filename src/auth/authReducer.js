import { types } from "../types/types"

export const authReducer = (state={}, action) => {



    switch (action?.type) {
        
        case types.login:
            return {
                // No entiendo por que utilizar el operador spread
                ...action.payload,
                logged: true
            }

        case types.logout:
            return {
                logged: false
            }
    
        default:
            return state
        }
}