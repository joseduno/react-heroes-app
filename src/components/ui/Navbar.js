import React, { useContext } from 'react'
import { Link, NavLink, useHistory } from 'react-router-dom'
import { AuthContext } from '../../auth/AuthContext'
import { types } from '../../types/types'

export const Navbar = () => {

    const {user, dispatch} = useContext(AuthContext)
    
    // useHistory: hook del BrowserRouter as Router, se puede utilizar por
    // componentes hijos (por ejemplo para redireccionar a otra ruta)
    const history = useHistory()

    // history.replace('/login') se utiliza para redireccionar 
    const handleLogout = () => {
        dispatch({
            type: types.logout
        })
        history.replace('/login')
    }

    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark container-fluid">

                <Link 
                    className="navbar-brand" 
                    to="/"
                >
                    Asociaciones
                </Link>

                
                    <div className="collapse navbar-collapse" id="navbarText">
                    <ul className="navbar-nav me-auto mb-2 mb-lg-0">

                        <NavLink 
                            activeClassName="active"
                            className="nav-item nav-link" 
                            exact
                            to="/marvel"
                        >
                            Marvel
                        </NavLink>

                        <NavLink 
                            activeClassName="active"
                            className="nav-item nav-link" 
                            exact
                            to="/dc"
                        >
                            DC
                        </NavLink>

                        <NavLink 
                            activeClassName="active"
                            className="nav-item nav-link" 
                            exact
                            to="/search"
                        >
                            Search
                        </NavLink>
                    </ul>

                    <span className="navbar-text text-info me-4">
                        {user.name}
                    </span>
                        <button
                            className="nav-item nav-link btn btn-primary"
                            onClick={handleLogout}
                        >
                            Logout
                        </button>
                </div>
        </nav>
    ) 
}