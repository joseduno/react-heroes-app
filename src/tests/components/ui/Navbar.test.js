import { mount } from "enzyme"
import { MemoryRouter, Router } from "react-router"
import { AuthContext } from "../../../auth/AuthContext"
import { Navbar } from "../../../components/ui/Navbar"
import { types } from "../../../types/types"


describe('Pruebas en el Navbar', () => {

    // Para emular el uso del useHistory, el componente que lo usa debe estar envuelto
    // en una etiqueta <Router history={historyMock}></Router> para que funcione la
    // prueba
    const historyMock = {
        location: {},
        replace: jest.fn(),
        listen: jest.fn(),
        createHref: jest.fn(),
    }

    const contextValue = {
        dispatch: jest.fn(),
        user: {logged: true, name: 'test'}
    }

    const wrapper = mount(
        <AuthContext.Provider value={contextValue} >
            <MemoryRouter>
                <Router history={historyMock}>
                    <Navbar />
                </Router>
            </MemoryRouter>
        </AuthContext.Provider>
    )

    afterEach(() => {
        jest.clearAllMocks()
    })

    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('.text-info').text().trim()).toBe('test')
    })

    test("debe llamar dispatch y usar el history.replace('/')", () => {
        wrapper.find('button').simulate('click')
        expect(contextValue.dispatch).toHaveBeenCalledWith({type: types.logout})
        expect(historyMock.replace).toHaveBeenCalledWith('/login')

    })
    
})
