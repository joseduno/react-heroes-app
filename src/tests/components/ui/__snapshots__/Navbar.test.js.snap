// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el Navbar debe mostrarse correctamente 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark container-fluid"
>
  <a
    className="navbar-brand"
    onClick={[Function]}
  >
    Asociaciones
  </a>
  <div
    className="collapse navbar-collapse"
    id="navbarText"
  >
    <ul
      className="navbar-nav me-auto mb-2 mb-lg-0"
    >
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        Search
      </a>
    </ul>
    <span
      className="navbar-text text-info me-4"
    >
      test
    </span>
    <button
      className="nav-item nav-link btn btn-primary"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
</nav>
`;
